src/parquet/arrow/reader.cc

```diff
@@ -1098,5 +1299,79 @@ Status ColumnReader::NextBatch(int batch_size, std::shared_ptr<Array>* out) {
   return impl_->NextBatch(batch_size, out);
 }
 
+// StructImpl methods
+
+Status StructImpl::CreateBitmap(size_t num_elements,
+    std::shared_ptr<PoolBuffer>* null_bitmap_out) {
+  *null_bitmap_out = std::make_shared<::arrow::PoolBuffer>(pool_);
+  auto bitmap_size = ::arrow::BitUtil::CeilByte(num_elements) / 8;
+  RETURN_NOT_OK((*null_bitmap_out)->Resize(bitmap_size));
+  uint8_t* null_bitmap_ptr = (*null_bitmap_out)->mutable_data();
+  memset(null_bitmap_ptr, 0xFF, bitmap_size);
+  return Status::OK();
+}
+
+Status StructImpl::DefLevelsToNullArray(
+    std::shared_ptr<PoolBuffer>* null_bitmap_out,
+    int64_t* null_count_out) {
+  std::shared_ptr<PoolBuffer> null_bitmap;
+  auto null_count = 0;
+  ValueLevelsPtr def_levels;
+  RETURN_NOT_OK(DefLevels(&def_levels));
+
+  RETURN_NOT_OK(CreateBitmap(def_levels->size(), &null_bitmap));
+  uint8_t* null_bitmap_ptr = null_bitmap->mutable_data();
+  for (uint i = 0; i < def_levels->size(); i++) {
+    if ((*def_levels)[i] < struct_def_level_) {
+      // Mark null
+      null_count += 1;
+      ::arrow::BitUtil::ClearBit(null_bitmap_ptr, i);
+    }
+  }
+
+  *null_count_out = null_count;
+  *null_bitmap_out = (null_count == 0) ? nullptr : null_bitmap;
+  return Status::OK();
+}
+
+Status StructImpl::DefLevels(ValueLevelsPtr* out) {
```

Comment:
Is caching needed here? For now, it looks like we only calculated this once. But it may be used multiple times.  I think we can add a comment about caching.